### PR TITLE
Analyzer added identifier typo corrections

### PR DESCRIPTION
**Pull request by kitaisreal**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

File: src/Analyzer/Passes/QueryAnalysisPass.cpp [hunk 1]
```diff
@@ -3485,6 +3653,9 @@ ProjectionNames QueryAnalyzer::resolveLambda(const QueryTreeNodePtr & lambda_nod
   */
 ProjectionNames QueryAnalyzer::resolveFunction(QueryTreeNodePtr & node, IdentifierResolveScope & scope)
 {
+    std::cout << "QueryAnalyzer::resolveFunction " << std::endl;
+    std::cout << node->dumpTree() << std::endl;
+
```
Comment (novikd) [inline src/Analyzer/Passes/QueryAnalysisPass.cpp:3658]: Forgot to remove it.

File: src/Analyzer/Passes/QueryAnalysisPass.cpp [hunk 2]
```diff
@@ -1352,6 +1375,140 @@ ProjectionName QueryAnalyzer::calculateSortColumnProjectionName(const QueryTreeN
...
+                column_identifier_with_table_name_and_database_name.push_back(column_identifier_part);
+
+            if (unresolved_identifier.getPartsSize() == column_identifier_with_table_name_and_database_name.getPartsSize())
+                valid_identifiers_result.insert(column_identifier_with_table_name_and_database_name);
+        }
+    }
+}
+
+/// Get valid identifiers for typo correction from scope without looking at parent scopes
+void QueryAnalyzer::getScopeValidIdentifiersForTypoCorrection(
```
Comment (novikd) [inline src/Analyzer/Passes/QueryAnalysisPass.cpp:1423]: Let's rename to `collectValidIdentifiersInScope`.

File: src/Analyzer/Passes/QueryAnalysisPass.cpp [hunk 3]
```diff
@@ -1352,6 +1375,140 @@ ProjectionName QueryAnalyzer::calculateSortColumnProjectionName(const QueryTreeN
...
+        {
+            getTableExpressionValidIdentifiersForTypoCorrection(unresolved_identifier,
+                table_expression,
+                table_expression_data,
+                valid_identifiers_result);
+        }
+
+    }
+
+    if (allow_function_identifiers && unresolved_identifier.isShort())
```
Comment (novikd) [inline src/Analyzer/Passes/QueryAnalysisPass.cpp:1449]: You check `unresolved_identifier.isShort()` several times. You can do it only once. Also, code to insert all the keys of the map to the vector can be extracted to a separate function and reused.

File: src/Analyzer/Passes/QueryAnalysisPass.cpp [hunk 4]
```diff
@@ -1352,6 +1375,140 @@ ProjectionName QueryAnalyzer::calculateSortColumnProjectionName(const QueryTreeN
     return sort_column_projection_name_buffer.str();
 }
 
+/// Get valid identifiers for typo correction from table expression
+void QueryAnalyzer::getTableExpressionValidIdentifiersForTypoCorrection(const Identifier & unresolved_identifier,
```
Comment (novikd) [inline src/Analyzer/Passes/QueryAnalysisPass.cpp:1379]: Rename to `collectValidIdentifiersInTableExpression`

File: tests/queries/0_stateless/02475_analyzer_subquery_compound_expression.sql [hunk 5]
```diff
@@ -0,0 +1,7 @@
+SET allow_experimental_analyzer=1;
+
+SELECT cast(tuple(1, 2), 'Tuple(value_1 UInt64, value_2 UInt64)') AS value, value.value_1, value.value_2;
+
+SELECT '--';
+
+SELECT value.value_1, value.value_2 FROM (SELECT cast(tuple(1, 2), 'Tuple(value_1 UInt64, value_2 UInt64)') AS value);
```
Comment (novikd) [inline tests/queries/0_stateless/02475_analyzer_subquery_compound_expression.sql:7]: The test doesn't seem to be related.

**Review by novikd — COMMENTED** [review]
In general OK

File: src/Analyzer/Passes/QueryAnalysisPass.cpp (hunk 1 above)
Comment (kitaisreal) [reply src/Analyzer/Passes/QueryAnalysisPass.cpp]: Updated.

**Comment by kitaisreal** [comment]
Fuzzer issues must be fixed by https://github.com/ClickHouse/ClickHouse/pull/43141, https://github.com/ClickHouse/ClickHouse/pull/43119, https://github.com/ClickHouse/ClickHouse/pull/43118.
